crates/matrix-sdk-base/src/read_receipts.rs

```diff
@@ -0,0 +1,395 @@
+use matrix_sdk_common::deserialized_responses::SyncTimelineEvent;
+use ruma::{
+    events::{
+        poll::{start::PollStartEventContent, unstable_start::UnstablePollStartEventContent},
+        receipt::{ReceiptThread, ReceiptType},
+        room::message::Relation,
+        AnySyncMessageLikeEvent, AnySyncTimelineEvent, OriginalSyncMessageLikeEvent,
+        SyncMessageLikeEvent,
+    },
+    serde::Raw,
+    EventId, OwnedEventId, RoomId, UserId,
+};
+use tracing::{field::display, instrument, trace};
+
+use super::BaseClient;
+use crate::{error::Result, store::StateChanges, RoomInfo};
+
+/// Provider for timeline events prior to the current sync.
+pub trait PreviousEventsProvider: Send + Sync {
+    /// Returns the list of known timeline events, in sync order, for the given
+    /// room.
+    // TODO: return a reference or some kind of iterator
+    fn for_room(&self, room_id: &RoomId) -> Vec<SyncTimelineEvent>;
```

Comment:
Reading the rest of the code and jumping back to here, I understand your `TODO` note. Indeed, returning `impl Iterator<Item = &SyncTimelineEvent>` would be nice.

Reply:
I've tried that, but can't return an impl in a trait, and I've tried a bit to make it an associated type or something like that and ran out of time. We can try that in a follow-up, but the current way is OK, we only have two impl, and one is the dummy one.